statistics/cmsketch.go

```diff
@@ -520,67 +468,103 @@ func (c *CMSketch) Copy() *CMSketch {
 	return &CMSketch{count: c.count, width: c.width, depth: c.depth, table: tbl, defaultValue: c.defaultValue}
 }
 
+// AppendTopN appends a topn into the TopN struct.
+func (c *TopN) AppendTopN(data []byte, count uint64) {
+	c.sorted = false
+	c.TopN = append(c.TopN, TopNMeta{data, count})
+}
+
+// GetWidthAndDepth returns the width and depth of CM Sketch.
+func (c *CMSketch) GetWidthAndDepth() (int32, int32) {
+	return c.width, c.depth
+}
+
+// CalcDefaultValForAnalyze calculate the default value for Analyze.
+// The value of it is count / NDV in CMSketch. This means count and NDV are not include topN.
+func (c *CMSketch) CalcDefaultValForAnalyze(NDV uint64) {
+	c.defaultValue = c.count / mathutil.MaxUint64(1, NDV)
+}
+
+// TopN stores most-common values, which is used to estimate point queries.
+type TopN struct {
+	TopN   []TopNMeta
+	sorted bool
+}
+
 // Copy makes a copy for current TopN.
 func (c *TopN) Copy() *TopN {
 	if c == nil {
 		return nil
 	}
-	topN := make(map[uint64][]*TopNMeta, len(c.topN))
-	for h1, vals := range c.topN {
-		newVals := make([]*TopNMeta, 0, len(vals))
-		for _, val := range vals {
-			newVal := TopNMeta{h2: val.h2, Count: val.Count, Data: make([]byte, len(val.Data))}
-			copy(newVal.Data, val.Data)
-			newVals = append(newVals, &newVal)
-		}
-		topN[h1] = newVals
-	}
-	return &TopN{topN}
-}
-
-// TopN gets all the topN meta.
-func (c *TopN) TopN() []*TopNMeta {
-	if c == nil {
-		return nil
+	topN := make([]TopNMeta, len(c.TopN))
+	for i, t := range c.TopN {
+		topN[i].Encoded = make([]byte, len(t.Encoded))
+		copy(topN[i].Encoded, t.Encoded)
+		topN[i].Count = t.Count
 	}
-	topN := make([]*TopNMeta, 0, len(c.topN))
-	for _, meta := range c.topN {
-		topN = append(topN, meta...)
+	return &TopN{
+		TopN:   topN,
+		sorted: c.sorted,
 	}
-	return topN
 }
 
-// TopNMap gets the origin topN map.
-func (c *TopN) TopNMap() map[uint64][]*TopNMeta {
-	return c.topN
+// TopNMeta stores the unit of the TopN.
+type TopNMeta struct {
+	Encoded []byte
+	Count   uint64
 }
 
-// AppendTopN appends a topn into the cm sketch.
-func (c *TopN) AppendTopN(data []byte, count uint64) {
-	if c.topN == nil {
-		c.topN = make(map[uint64][]*TopNMeta)
+// QueryTopN returns the results for (h1, h2) in murmur3.Sum128(), if not exists, return (0, false).
+func (c *TopN) QueryTopN(d []byte) (uint64, bool) {
+	if c == nil {
+		return 0, false
+	}
+	idx := c.findTopN(d)
+	if idx < 0 {
+		return 0, false
 	}
-	h1, h2 := murmur3.Sum128(data)
-	c.topN[h1] = append(c.topN[h1], &TopNMeta{h2, data, count})
+	return c.TopN[idx].Count, true
 }
 
-// GetWidthAndDepth returns the width and depth of CM Sketch.
-func (c *CMSketch) GetWidthAndDepth() (int32, int32) {
-	return c.width, c.depth
+func (c *TopN) findTopN(d []byte) int {
+	if c.sorted == false {
+		c.Sort()
+	}
```

Comment:
It's better to sort it after building. Sort it when querying may cause data race.